core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
           var isStreamCopied: Boolean = false
           try {
             input = streamWrapper(blockId, in)
-            // Only copy the stream if it's wrapped by compression or encryption, also the size of
-            // block is small (the decompressed block is smaller than maxBytesInFlight)
-            if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {
+            // Only copy the stream if it's wrapped by compression or encryption upto a size of
+            // maxBytesInFlight/3. If stream is longer, then corruption will be caught while reading
+            // the stream.
+            streamCompressedOrEncrypted = !input.eq(in)
+            if (detectCorrupt && streamCompressedOrEncrypted) {
               isStreamCopied = true
               val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
-              // Decompress the whole block at once to detect any corruption, which could increase
-              // the memory usage tne potential increase the chance of OOM.
+              // Decompress the block upto maxBytesInFlight/3 at once to detect any corruption which
+              // could increase the memory usage and potentially increase the chance of OOM.
               // TODO: manage the memory used here, and spill it into disk in case of OOM.
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              isStreamCopied = Utils.copyStreamUpTo(
+                input, out, maxBytesInFlight / 3, closeStreams = true)
+              if (isStreamCopied) {
+                input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              } else {
+                input = new SequenceInputStream(
+                  out.toChunkedByteBuffer.toInputStream(dispose = true), input)
+              }
```

Comment:
I think it would be better if you refactored this so `copyStreamUpTo` includes this part as well, and just returns an InputStream.  The main reason is for testing -- I worry about edge cases here.   If this is factored out, it would be much easier to test that the resulting input stream is OK if eg. there is an empty input, or just up to the limit, or over the limit by X, and make sure streams get closed correctly etc.